dotnet/src/SemanticKernel/KernelBuilder.cs

```diff
@@ -89,7 +89,19 @@ public KernelBuilder WithLogger(ILogger log)
     public KernelBuilder WithMemory(ISemanticTextMemory memory)
     {
         Verify.NotNull(memory);
-        this._memory = memory;
+        this._memoryFactory = () => memory;
+        return this;
+    }
+
+    /// <summary>
+    /// Add a semantic text memory store factory.
+    /// </summary>
+    /// <param name="factory">The store factory.</param>
+    /// <returns>Updated kernel builder including the semantic text memory entity.</returns>
+    public KernelBuilder WithMemory<TStore>(Func<(ILogger Logger, KernelConfig Config), TStore> factory) where TStore : ISemanticTextMemory
```

Comment:
Might be nice to update comments to help users. I still personally get confused on the difference between ISemanticTextMemory and IMemoryStore and when I would chose one or the other and whether I should ever need to add both. I know there are TODOs here regarding `UseMemory` so can be resolved and dealt with separately.